### PR TITLE
[Associated type inference] Various improvements to eliminate crashes and infer more often

### PR DESCRIPTION
**Pull request by DougGregor**
Fix several related issues with associated type inference:

* We had a path where we would record a type witness that contained an error but never report said error, leading to downstream crashes (e.g, the crash in [SR-6609](https://bugs.swift.org/browse/SR-6609) / rdar://problem/36038033).
* We could end up recursing infinitely when looking up type witnesses, so block said recursion appropriately (fixed another two crashers in our suite).
* Don't allow typealiases in protocol extensions to block associated type inference when they don't match, which fixes associated type inference for [SR-6609](https://bugs.swift.org/browse/SR-6609) / rdar://problem/36038033 for real.
* When substituting with tentative type witnesses, allow us to find tentative type witnesses within the same protocol hierarchy. Fixes rdar://problem/36453271.
* Somewhere along the way, we fixed rdar://problem/16316115, so pull in that test case. That bug is from 3/13/14, in case anyone cares :)

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

File: lib/Sema/TypeCheckProtocol.cpp [hunk 1]
```diff
@@ -2746,6 +2750,16 @@ ResolveWitnessResult ConformanceChecker::resolveTypeWitnessViaLookup(
     }
   }
 
+  // If there are no viable witnesses, and all nonviable candidates came from
+  // protocol extensions, treat this as "missing".
+  if (viable.empty() &&
+      std::find_if(nonViable.begin(), nonViable.end(),
+                   [](const std::pair<TypeDecl *, CheckTypeWitnessResult> &x) {
+                     return x.first->getDeclContext()
+                        ->getAsProtocolOrProtocolExtensionContext() == nullptr;
```
Comment (xedin) [inline lib/Sema/TypeCheckProtocol.cpp]: The comment at the top mentioned "protocol extensions", so I'm wondering is it _only_ protocol extensions? If so I think we should be using getAsProtocolExtensionContext() ?

**Review by xedin — APPROVED** [review]
LGTM!

File: lib/Sema/TypeCheckProtocol.cpp (hunk 1 above)
Comment (DougGregor) [reply lib/Sema/TypeCheckProtocol.cpp]: Typealiases in protocols would be okay here, too; if we see them along this path, it's fine to treat them the same way as typealiases in protocol extensions.

**Comment by DougGregor** [comment]
Thanks @xedin

File: lib/Sema/TypeCheckProtocol.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/TypeCheckProtocol.cpp]: Gotcha, thanks!

**Pull request merged**